ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/BlockchainQueries.java

```diff
@@ -87,6 +96,24 @@ public BlockchainQueries(
             ? Optional.of(
                 new TransactionLogBloomCacher(blockchain, cachePath.get(), scheduler.get()))
             : Optional.empty();
+
+    blockBodySoftCache =
+        CacheBuilder.newBuilder()
+            .maximumSize(BLOCK_BODY_SOFT_CACHE_SIZE)
+            .softValues()
+            .build(CacheLoader.from(blockchain::getBlockBody));
+
+    blockchain.observeBlockAdded(this::onBlockAdded);
+    blockchain.observeChainReorg(this::onChainReorg);
+  }
+
+  private void onChainReorg(
+      final BlockWithReceipts blockWithReceipts, final Blockchain blockchain) {
+    blockBodySoftCache.invalidate(blockWithReceipts.getHash());
+  }
```

Comment:
By my reading, `onChainReorg` gives you the blocks on the fork, why would we want to invalidate them?

Reply:
In case we had a "cache miss" where the block body for that new head was empty.  Are we guaranteed to see that head in blockAdded as well?  If so this can be removed.

Reply:
No, it's good as it is. The block added event only happens once and not for each block on the new fork. Interesting question as to if it should though...